xla/service/algebraic_simplifier.cc

```diff
@@ -8077,6 +8077,33 @@ absl::Status AlgebraicSimplifierVisitor::HandleSelect(HloInstruction* select) {
                       select->mutable_operand(0)->shape(), HloOpcode::kNot,
                       select->mutable_operand(0)));
     }
+    // select(compare(a, b, GT/GE), a, b) => or(a, b)
+    // select(compare(a, b, LT/LE), a, b) => and(a, b)
+    // select(compare(a, b, EQ), a, b) => b
+    // select(compare(a, b, NE), a, b) => a
+    HloInstruction *lhs, *rhs, *compare;
+    if (Match(select, m::Select(m::Compare(&compare, m::Op(&lhs), m::Op(&rhs)),
+                                m::Op(&lhs), m::Op(&rhs)))) {
```

Comment:
I think this usage of the matcher does not make sure that the first operand of the compare is the same as the second operand of the select. Same for second operand of compare/third operand of select. This needs to be tested separately. Please add a test case where this would be caught.

Reply:
Good catch! Fixed. 
Now I call `Match()` two times - for `select` and for `compare` ops separately. The second Match call uses `m::Op().Is(lhs), m::Op().Is(rhs)` to make sure the operands order in `compare` Op is the same as in `select` Op.